fix(feature:path-tracking, core:network): fix bugs in path tracker fe…

sam-arth07 opened the pull request:
Fixes - [Jira-#495](https://mifosforge.jira.com/browse/MIFOSAC-495)

Please Add Screenshots If there are any UI changes.

Before:

<img src="https://github.com/user-attachments/assets/ed2cb8ca-a0cc-4844-b128-60240a947ced" width="240"/>

After: 

https://github.com/user-attachments/assets/74dcb4a7-c3b6-4f42-a4a1-af243913ff24 
 
### Changes made in the PR: 
- Included changes related to `location permission handling and correct package name` already done in this draft pr: https://github.com/openMF/android-client/pull/2437 
- resolved 404 network calls by creating a new datatable `m_staff_path_tracking ` with appropriate fields and updated the `DataTableService` to use this new datatable endpoint. 
- Added Logic to display accurate start, stop, and route markers in `PathTrackingScreen.android.kt`.
- Validated that timestamps match the actual tracking session and multiple sessions display correctly.
- Verified that the functionality is working on different network conditions (4G (LTE), 3G (UMTS) , 2G (EDGE), WiFi).

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the static analysis check `./gradlew check` or `ci-prepush.sh` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them.

sam-arth07 committed:
fix(feature:path-tracking, core:network): fix bugs in path tracker feature.

therajanmaurya commented on core/model/src/commonMain/kotlin/com/mifos/core/model/objects/users/UserLocation.kt:
```diff
@@ -11,10 +11,12 @@ package com.mifos.core.model.objects.users
 /**
  * Created by Rajan Maurya on 24/01/17.
  */
+@Serializable
```
We are no longer using Serializable, @niyajali confirm?

HekmatullahAmin replied:
@therajanmaurya i belive for now we are no longer using parcelizable cause it is android specific, and in ios it doesn't have some annotations

HekmatullahAmin commented on .gitignore:
```diff
@@ -12,4 +12,5 @@ mifosng-android/mifosng-android.iml
-.kotlin/
\ No newline at end of file
+.kotlin/
+build-logic/convention/bin
```
use `bin/` so all of them git ignored

HekmatullahAmin commented on feature/path-tracking/src/androidMain/kotlin/com/mifos/feature/path/tracking/PathTrackingScreen.android.kt:
```diff
@@ -127,35 +124,36 @@ actual fun PathTrackingMapView(latLngList: List<UserLatLng>) {
-    )
-}
+    ) {
+        if (latLngList.isNotEmpty()) {
```
use proper name for `latLngList` it is not understandable. maybe use full form instead of abbreviation

biplab1 replied:
Just a note — LatLng is actually the standard term used in the Google Maps API, so using latLngList here is appropriate and widely understood in that context.

Example:
`position = CameraPosition.fromLatLngZoom(LatLng(latLng.lat, latLng.lng), 15f)`

biplab1 commented on core/network/src/commonMain/kotlin/com/mifos/core/network/datamanager/DataManagerDataTable.kt:
```diff
@@ -71,7 +71,7 @@ class DataManagerDataTable(
 
     /**
-     * This Method is adding the User Tracking Data in the data Table "user_tracking"
+     * This Method is adding the User Tracking Data in the data Table "client_pinpoint_location"
```
It should be `m_staff_path_tracking` instead.

biplab1 commented on core/network/src/commonMain/kotlin/com/mifos/core/network/datamanager/DataManagerDataTable.kt:
```diff
@@ -85,7 +85,7 @@ class DataManagerDataTable(
 
     /**
-     * This Method is fetching the User Path Tracking from the DataTable "user_tracking"
+     * This Method is fetching the User Path Tracking from the DataTable "client_pinpoint_location"
```
Same here, please review

biplab1 reviewed (commented):
Thanks for the demo! I have a few questions for better clarity:
1. Could you share a screenshot of the backend data that was captured?
2. Which entity was used to create the DataTable used for path tracking (e.g., client, group, center, office)?
3. Where in the `web-app` can we view the DataTable values for the user shown in the screen recording?
4. What do the start and end positions represent? Are they the actual initial and final locations of the trip? Similarly, are the start time and end time automatically recorded timestamps, or are they estimated durations between manually selected points?
>Just to clarify — the start and end locations should be captured automatically when the journey begins and ends, not based on estimated locations selected by the field officer on the map.
5. The field officer shouldn’t be able to edit start/end locations manually — these should come from live GPS data once tracking begins.
6. How is path tracking started or stopped?
7. Was this tested on a physical device or an emulator?
    (For location-based features, physical devices are preferred as emulators may return default or inaccurate GPS data.)
8. When a field officer starts path tracking and stops it upon reaching the destination, a new card should appear in the UI, and the corresponding data should be saved to the DataTable.
9. Edge cases should be handled — for example, if the field officer starts and stops tracking at nearly the same location (within a small epsilon distance), then that entry should be ignored and not recorded.

sam-arth07 committed:
Merge branch 'kmp-impl' into MIFOSAC-495

biplab1 commented:
Here are my thoughts on your responses and the current implementation:

1. I was referring to the data captured at the backend, not the logcat output. From your screen recording (point 3), it appears that no actual data is being captured — only dummy values are present in the `DataTable`.

2. Path tracking data shouldn't be linked to the `client` entity, as it's related to `staff` (field officers). A staff member may or may not be a client of the organization.
   A better approach would be to associate the path tracking `DataTable` with the `office` entity, since staff are linked to offices. You can save both `officeId` and `staffId` in the table.
   To create a staff member, refer to:
   * [Manage Employees](https://docs.mifos.org/mifosx/user-manual/for-administrators-mifos-x-platform/administration/organization/manage-employees)
   * [Create Users](https://docs.mifos.org/mifosx/user-manual/for-administrators-mifos-x-platform/administration/users)
   
     After creating an employee, you'll also need to assign them as a user.

3. Path tracking data should not appear on the client page — there is no such concept as `client` path tracking. Please remove the table from the DataTables section.
   Also, it looks like you're currently fetching the path tracking `DataTable` using `clientId` instead of the ID of the currently logged-in `user`.

4. This indicates that path tracking was never properly implemented. The original version may have been using mocked data as well.

5. Can be handled in the same PR as in point 6

6. It's your choice whether to address these issues in this PR or in a follow-up PR.

7. Thanks for confirming. Good to know the recording is from a physical device — that's important for reliable GPS testing, since emulators can return default values.

8 & 9. You can add the following improvements in your next PR:

* When a field officer starts and then stops path tracking, a new card should appear in the UI, and the values should be stored in the `DataTable`.
* Handle edge cases where tracking starts and stops at nearly the same location (within some epsilon distance) — in such cases, the entry should be ignored to avoid unnecessary records.

10. Optional Suggestion:
If feasible, path tracking could show the actual route taken, similar to how apps like Uber or Ola display ride history — instead of just a straight line between the start and end locations.

     Alternatively, if the goal is only to capture start and end points, a separate user flow could be introduced. In that flow, the app would record only the GPS coordinates at the beginning and end of the journey, manually — this would significantly reduce battery usage by avoiding continuous location tracking. However, this raises an important question:
     > Would this still be considered "path tracking," or simply a start/end location record?

sam-arth07 committed:
Merge branch 'kmp-impl' into MIFOSAC-495

sam-arth07 committed:
Merge branch 'kmp-impl' into MIFOSAC-495

sam-arth07 committed:
fix: moved m_staff_path_tracking datatable from client to staff, verfy view path_tracking datatable functionality.

sam-arth07 commented:
@HekmatullahAmin @niyajali I have made the requested changes, if there is anything else you can let me know, I will try to make changes as earliest as possible, else it is ready to be merged.

Also @biplab1, I have changed the datatable from prev being linked to client entity to a office entity now which is then assigned to a user as well, also deleted the `m_staff_path_tracking` datatable from the client datatables section. 

Attached screenshots for reference:

<img width="1911" height="950" alt="image" src="https://github.com/user-attachments/assets/434463c8-7b20-48b1-9d1c-fcb50c268245" />

<img width="1684" height="785" alt="image" src="https://github.com/user-attachments/assets/80a7a9ac-0fbe-4372-989f-60019fd2d740" />

As for the start path tracking feature, i had a discussion with @therajanmaurya , he mentioned that it could be taken up in future PRs. For this PR, only being able to display the mapview on the Path Tracking Screen is enough.
Thank you for your suggestions and guidance.

sam-arth07 committed:
fix(feature/path-tracking): added logic to extract location from latLng

sam-arth07 committed:
fix: static checks pass
